master/internal/api_trials.go

```diff
@@ -93,6 +93,61 @@ type TrialLogBackend interface {
 	DeleteTrialLogs(trialIDs []int) error
 }
 
+// Catches information on active running trials.
+type trialAllocation struct {
+	Pulling  bool
+	Running  bool
+	Starting bool
+	Task     model.TaskID
+}
+
+func (a *apiServer) enrichTrialState(trials []*trialv1.Trial) ([]*trialv1.Trial, error) {
+	// filter allocations by TaskIDs on this page of trials
+	taskFilter := make([]string, 0, len(trials))
+	for _, trial := range trials {
+		taskFilter = append(taskFilter, trial.TaskId)
+	}
+
+	// get active trials by TaskId
+	tasks := []trialAllocation{}
+	err := a.m.db.Query(
+		"get_active_allocations_by_task",
+		&tasks,
+		strings.Join(taskFilter, ","),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Collect state information by TaskID
+	byTaskID := make(map[model.TaskID]experimentv1.State, len(tasks))
+	for _, task := range tasks {
+		switch {
+		case task.Running:
+			byTaskID[task.Task] = experimentv1.State_STATE_RUNNING
+		case task.Starting:
+			byTaskID[task.Task] = experimentv1.State_STATE_STARTING
+		case task.Pulling:
+			byTaskID[task.Task] = experimentv1.State_STATE_PULLING
+		default:
+			byTaskID[task.Task] = experimentv1.State_STATE_QUEUED
+		}
+	}
+
+	// Active trials converted to Queued, Pulling, Starting, or Running
+	for _, trial := range trials {
+		if trial.State == experimentv1.State_STATE_ACTIVE {
+			if setState, ok := byTaskID[model.TaskID(trial.TaskId)]; ok {
+				trial.State = setState
+			} else {
+				trial.State = experimentv1.State_STATE_QUEUED
+			}
+		}
+	}
```

Comment:
if you wanted to make this logic generic, probably the easiest approach would be to make `trialAllocation`/`experimentAllocation` implement some `aggregateAllocationStateWithID` interface and then make this a func that takes that.